File: simple-statemachine-with-persistence/using-statemachine-recipe/simple-statemachine-java/src/main/java/com/example/statemachine/simple/doument/DocumentStateMachineRestController.java
package com.example.statemachine.simple.doument;

import com.example.statemachine.simple.doument.config.DocumentIdChecker;
import com.example.statemachine.simple.doument.config.exception.PersistenceDataIsNotExistException;
import com.example.statemachine.simple.doument.persist.DocumentStateMachine;
import java.util.List;
import java.util.Map;
import org.springframework.http.ResponseEntity;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RestController;

@RestController
public class DocumentStateMachineRestController {

  private final JdbcTemplate jdbcTemplate;
  private final DocumentIdChecker documentIdChecker;
  private final DocumentStateMachine documentStateMachine;

  public DocumentStateMachineRestController(
    final JdbcTemplate jdbcTemplate,
    final DocumentStateMachine documentStateMachine) {
    this.jdbcTemplate = jdbcTemplate;
    this.documentIdChecker = new DocumentIdChecker(jdbcTemplate);
    this.documentStateMachine = documentStateMachine;
  }

  @GetMapping("/state-list")
  public DocumentState[] getStateList() {

    return DocumentState.values();
  }


  @GetMapping("/event-list")
  public DocumentEvent[] getEventList() {

    return DocumentEvent.values();
  }

  @GetMapping("/document-list")
  public List<DocumentVo> documentList() {

    return documentStateMachine.listDbEntries();
  }


  @GetMapping("/{documentId}/current-state")
  public DocumentState currentState(@PathVariable("documentId") final int documentId) {

    return getStateFromDatabase(documentId);
  }


  @PostMapping("/{documentId}/send/{event}")
  public EventResultStatement sendEvent(
    @PathVariable("documentId") final int documentId,
    @PathVariable("event") final DocumentEvent event) {

    final var prevState = getStateFromDatabase(documentId);

    documentStateMachine.change(documentId, event);

    final var changedState = getStateFromDatabase(documentId);

    return new EventResultStatement(documentId, event, new State(prevState, changedState));
  }

  @ExceptionHandler(PersistenceDataIsNotExistException.class)
  public ResponseEntity<Map<String, String>> handlePersistenceDataIsNotExistException(
    final PersistenceDataIsNotExistException ex) {

    return ResponseEntity.unprocessableEntity()
      .body(Map.of("message", ex.getMessage()));
  }

  private DocumentState getStateFromDatabase(final int documentId)
    throws PersistenceDataIsNotExistException {

    documentIdChecker.isExist(documentId);

    return jdbcTemplate.queryForObject(
      "SELECT state FROM documents WHERE id = ?",
      DocumentState.class, documentId);
  }

  public record EventResultStatement(int documentId, DocumentEvent event, State state) {

  }

  public record State(DocumentState prev, DocumentState changed) {

  }
}

File: simple-statemachine-with-persistence/using-statemachine-recipe/simple-statemachine-kotlin/src/main/kotlin/com/example/statemachine/simple/document/DocumentStateMachineRestController.kt
package com.example.statemachine.simple.document

import com.example.statemachine.simple.document.config.DocumentIdChecker
import com.example.statemachine.simple.document.config.exception.PersistenceDataIsNotExistException
import com.example.statemachine.simple.document.persist.DocumentStateMachine
import org.springframework.http.ResponseEntity
import org.springframework.jdbc.core.JdbcTemplate
import org.springframework.web.bind.annotation.*

@RestController
class DocumentStateMachineRestController(
  private val jdbcTemplate: JdbcTemplate,
  private val documentStateMachine: DocumentStateMachine
) {
  private val documentIdChecker: DocumentIdChecker = DocumentIdChecker(jdbcTemplate)


  @get:GetMapping("/state-list")
  val stateList: Array<DocumentState>
    get() = DocumentState.entries.toTypedArray()

  @get:GetMapping("/event-list")
  val eventList: Array<DocumentEvent>
    get() = DocumentEvent.entries.toTypedArray()

  @GetMapping("/document-list")
  fun documentList(): List<DocumentVo> {
    return documentStateMachine.listDbEntries()
  }

  @GetMapping("/{documentId}/current-state")
  fun currentState(@PathVariable("documentId") documentId: Int): DocumentState {
    return getStateFromDatabase(documentId)
  }

  @PostMapping("/{documentId}/send/{event}")
  fun sendEvent(
    @PathVariable("documentId") documentId: Int,
    @PathVariable("event") event: DocumentEvent
  ): EventResultStatement {
    val prevState = getStateFromDatabase(documentId)

    documentStateMachine.change(documentId, event)

    val changedState = getStateFromDatabase(documentId)

    return EventResultStatement(documentId, event, State(prevState, changedState))
  }


  @ExceptionHandler(PersistenceDataIsNotExistException::class)
  fun handlePersistenceDataIsNotExistException(
    ex: PersistenceDataIsNotExistException
  ): ResponseEntity<Map<String, String>> {
    return ResponseEntity.unprocessableEntity()
      .body(mapOf("message" to ex.message!!))
  }


  @Throws(PersistenceDataIsNotExistException::class)
  private fun getStateFromDatabase(documentId: Int): DocumentState {
    documentIdChecker.isExist(documentId)
    return jdbcTemplate.queryForObject(
      "SELECT state FROM documents WHERE id = ?",
      DocumentState::class.java, documentId
    )
  }

  data class State(val prev: DocumentState, val changed: DocumentState)

  data class EventResultStatement(val documentId: Int, val event: DocumentEvent, val state: State)
}
